concordium-consensus/src/Concordium/KonsensusV1/Consensus/Quorum.hs

```diff
@@ -0,0 +1,238 @@
+                            RecentBlock (BlockPending _) ->
+                                return Awaiting
+                            -- The block is executed but not finalized.
+                            -- Perform the remaining checks before processing the 'QuorumMessage'.
+                            RecentBlock (BlockAlive quorumMessagePointer)
+                                -- Inconsistent rounds of the quorum signature message and the block it points to.
+                                | blockRound quorumMessagePointer /= qmRound -> do
+                                    flag $
+                                        RoundInconsistency
+                                            qmFinalizerIndex
+                                            qmRound
+                                            (blockRound quorumMessagePointer)
+                                    return Rejected
+                                -- Inconsistent epochs of the quorum signature message and the block it points to.
+                                | blockEpoch quorumMessagePointer /= qmEpoch -> do
+                                    flag $
+                                        EpochInconsistency
+                                            qmFinalizerIndex
+                                            qmEpoch
+                                            (blockEpoch quorumMessagePointer)
+                                    return Rejected
+                                -- Store, relay and process the quorum message.
+                                | otherwise -> do
+                                    storeQuorumMessage qm finalizerWeight
+                                    return Received
+    -- Extract the quorum signature message
+    getQuorumSignatureMessage skovData =
+        let genesisHash = skovData ^. genesisMetadata . to gmFirstGenesisHash
+        in  quorumSignatureMessageFor qm genesisHash
+    -- Get the finalizer if it is present in the current finalization committee otherwise return 'Nothing'.
+    getFinalizerByIndex skovData =
+        let finalizersForCurrentEpoch = skovData ^. skovEpochBakers . currentEpochBakers . bfFinalizers
+        in  finalizerByIndex finalizersForCurrentEpoch qmFinalizerIndex
+    -- Checks whether a finalizer has signed off a quorum message already.
+    hasSignedOffQuorumMessage finalizerIndex skovData =
+        isJust $! skovData ^? currentQuorumMessages . smFinalizerToQuorumMessage . ix finalizerIndex
+    -- Checks whether a finalizer has signed off a timeout mesage already.
+    hasSignedOffTimeoutMessage = False -- TODO: check in the timeout messages for the current round also.
+    -- Checks if the finalizer has signed off either a quorum or a timeout message already.
+    isDoubleSigning finalizerIndex skovData =
+        hasSignedOffQuorumMessage finalizerIndex skovData || hasSignedOffTimeoutMessage
```

Comment:
Double signing should only count if two *different* blocks were signed by quorum messages. It is fine for a finalizer to sign a block and a timeout message.